src/command/lsp/mod.rs

```diff
@@ -0,0 +1,230 @@
+
+        while let Some(event) = stream.next().await {
+            match event {
+                CompositionEvent::Started => {
+                    language_server.composition_did_start().await;
+                }
+                CompositionEvent::Success(CompositionSuccess {
+                    supergraph_sdl,
+                    federation_version,
+                    hints,
+                }) => {
+                    debug!("Successfully composed with version {}", federation_version);
+                    // Clear any previous errors on `supergraph.yaml`
+                    language_server
+                        .publish_diagnostics(supergraph_yaml_url.clone(), vec![])
+                        .await;
+                    language_server
+                        .composition_did_update(
+                            Some(supergraph_sdl),
+                            hints.into_iter().map(Into::into).collect(),
+                            None,
+                        )
+                        .await;
+                }
+                CompositionEvent::Error(CompositionError::Build { source: errors }) => {
+                    debug!(
+                        ?errors,
+                        "Composition {federation_version} completed with errors"
+                    );
+                    // Clear any previous errors on `supergraph.yaml`
+                    language_server
+                        .publish_diagnostics(supergraph_yaml_url.clone(), vec![])
+                        .await;
+                    language_server
+                        .composition_did_update(
+                            None,
+                            errors.into_iter().map(Into::into).collect(),
+                            None,
+                        )
+                        .await
+                }
+                CompositionEvent::Error(err) => {
+                    debug!("Composition {federation_version} failed: {err}");
+                    // TODO: we could highlight the version of federation, since it failed.
+                    let message = format!("Failed run composition {federation_version}: {err}",);
+                    let diagnostic = Diagnostic::new_simple(Range::default(), message);
+                    language_server
+                        .publish_diagnostics(supergraph_yaml_url.clone(), vec![diagnostic])
+                        .await;
+                }
+                CompositionEvent::SubgraphAdded(CompositionSubgraphAdded {
+                    name,
+                    schema_source,
+                }) => {
+                    debug!("Subgraph {} added", name);
+                    language_server.add_subgraph(name, schema_source).await;
+                }
+                CompositionEvent::SubgraphRemoved(CompositionSubgraphRemoved { name }) => {
+                    debug!("Subgraph {} removed", name);
+                    language_server.remove_subgraph(&name).await;
+                }
```

Comment:
These are net new events because the language server needs to know when subgraphs are added and removed, so I expanded out the definition of CompositionEvents to include these.

Reply:
trying to get an initial review out, but self-note: probs need to check that all the places we add/remove subgraphs are emitting the right event